tools/witx/src/abi.rs

```diff
@@ -227,91 +276,358 @@ def_instruction! {
+        /// representation as an `i64`.
+        I64FromBitflags {
+            ty: &'a RecordDatatype,
+            name: &'a NamedType,
+        } : [1] => [1],
+        /// Converts a native wasm `i32` to a language-specific record-of-bools.
+        BitflagsFromI32 {
+            repr: IntRepr,
+            ty: &'a RecordDatatype,
+            name: &'a NamedType,
+        } : [1] => [1],
+        /// Converts a native wasm `i64` to a language-specific record-of-bools.
+        BitflagsFromI64 {
+            repr: IntRepr,
+            ty: &'a RecordDatatype,
+            name: &'a NamedType,
+        } : [1] => [1],
+
+        // variants
+
+        /// This is a special instruction used at the entry of blocks used as
+        /// part of `ResultLower`, representing that the payload of that variant
+        /// being matched on should be pushed onto the stack.
+        VariantPayload : [0] => [1],
+
+        /// Pops a variant off the stack as well as `ty.cases.len()` blocks
+        /// from the code generator. Uses each of those blocks and the value
+        /// from the stack to produce `nresults` of items.
+        VariantLower {
+            ty: &'a Variant,
+            name: Option<&'a NamedType>,
+            nresults: usize,
+        } : [1] => [*nresults],
+
+        /// Pops an `i32` off the stack as well as `ty.cases.len()` blocks
+        /// from the code generator. Uses each of those blocks and the value
+        /// from the stack to produce a final variant.
+        VariantLift {
+            ty: &'a Variant,
+            name: Option<&'a NamedType>,
+        } : [1] => [1],
+
+        // calling/control flow
+
+        /// Represents a call to a raw WebAssembly API. The module/name are
+        /// provided inline as well as the types if necessary.
+        CallWasm {
+            module: &'a str,
+            name: &'a str,
+            params: &'a [WasmType],
+            results: &'a [WasmType],
+        } : [params.len()] => [results.len()],
+
+        /// Same as `CallWasm`, except the dual where an interface is being
+        /// called rather than a raw wasm function.
+        CallInterface {
+            module: &'a str,
+            func: &'a Function,
+        } : [func.params.len()] => [func.results.len()],
+
+        /// Returns `amt` values on the stack. This is always the last
+        /// instruction.
+        Return { amt: usize } : [*amt] => [0],
+
+        // ...
+
+        /// An instruction from an extended instruction set that's specific to
+        /// `*.witx` and the "Preview1" ABI.
+        Witx {
+            instr: &'a WitxInstruction<'a>,
+        } : [instr.operands_len()] => [instr.results_len()],
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Bitcast {
+    // Upcasts
+    F32ToF64,
+    F32ToI32,
+    F64ToI64,
+    I32ToI64,
+    F32ToI64,
+
+    // Downcasts
+    F64ToF32,
+    I32ToF32,
+    I64ToF64,
+    I64ToI32,
+    I64ToF32,
+
+    None,
+}
```

Comment:
Bitcasts between types with different sizes are sensitive to endianness. For example, in `F32ToF64`, does the `F32` go in the most significant half of the `F64` or the least significant half? Cross-endian configurations are a theoretical concern at this point, but I think we could at least document what should happen. Since wasm itself is little-endian, I propose this say "bitcasts between types with different sizes use little-endian byte ordering".

Also, it'd be good to mention here that widening bitcasts zero-extend.

Reply:
Oh I've figured that like wasm everything is little-endian here. For conversions like f32 to f64 I'm imagining it's the same as `f64::from(1.0f32)` in Rust where it's not really about moving bits but the f64 value losslessly matches the f32 value.

But yeah I'll definitely clarify this and indicate that everything is zero-extended. I haven't thought too too hard about the semantics here, but I think this'll all be ok.

Reply:
Ah. `f64::from` is lossless except it converts signaling NaN to quiet NaN, which is obscure, but surprising given that the rest of the ABI preserves NaN bit patterns. But this isn't urgent to sort out now.